[Merged by Bors] - feat(linear_algebra/tensor_algebra): Tensor algebras

adamtopaz opened the pull request:
This PR constructs the tensor algebra of a module over a commutative ring.

The main components are:
1. The construction of the tensor algebra: `tensor_algebra R M` for a module `M` over a commutative ring `R`.
2. The linear map `univ R M` from `M` to `tensor_algebra R M`.
3. Given a linear map `f`from `M`to an `R`-algebra `A`, the lift of `f` to `tensor_algebra R M` is denoted `lift R M f`.
4. A theorem `univ_comp_lift` asserting that the composition of `univ R M` with `lift R M f`is `f`.
5. A theorem `lift_unique` asserting the uniqueness of `lift R M f`with respect to property 4.

---
<!-- put comments you want to keep out of the PR commit here -->

Note: This PR only does the bare minimum!

adamtopaz committed:
Add defn of tensor algebra + univ prop.

adamtopaz committed:
Some linting

adamtopaz committed:
Add minor documentation.

adamtopaz committed:
Add header

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,378 @@
+-- force of to be linear
+| add_lin {a b : M} : rel ↑(a + b) (↑a + ↑b)
+| smul_lin {r : R} {a : M} : rel ↑(r • a) (r • ↑a)
+| zero_lin : rel ↑(0 : M) 0
```
It would be neat if there was some way to reuse `is_linear_map` or `linear_map` here.

On that note, `zero_lin` is unecessary as you can get it from simplifying `@smul_lin 0 _` I think

adamtopaz replied:
I don't think you can get by with only `smul_lin` since the `0` from `M` a priori has nothing to do with `pre.zero`.

It's possible there's some way to combine this with some of the other relations, but I feel like a small bit of redundancy is fine if it makes the construction more understandable.

eric-wieser replied:
Could you eliminate `pre.zero` and replace it with `pre.of 0`? or does that make things harder below?

adamtopaz replied:
That's possible. The thing is, I can imagine in the future having a more general construction which constructs rings given certain input data/relations, and this being a special case. This is mainly why I wanted to explicitly include all of the usual constructors for an algebra in pre, so that it would be easier to generalize later.

adamtopaz replied:
On second thought, I think it would make sense to change the constructors to something like this:
```lean
inductive pre
| of : M → pre
| const : R → pre
| mul : pre → pre → pre
| add : pre → pre → pre
```
I'm making this change now.

eric-wieser replied:
`of_scalar` might be a better name than `const`

eric-wieser replied:
That construction looks harder to prove `has_scalar` in though.

adamtopaz replied:
Not really. It's just multiplication by something of the form `of_scalar r`

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,378 @@
+/--
+The tensor algebra of the module `M` over the commutative ring `R`.
+-/
+def tensor_algebra := quot (tensor_algebra.rel R M)
```
`tensor_product` uses something more like `(add_con_gen (tensor_algebra.rel R M)).quotient` here - what's the tradeoff between the two approaches?

adamtopaz replied:
As I understand it, `add_con_gen` is essentially used in an additive context to generate, given `rel`, the smallest equivalence relation containing `rel` and which is compatible with addition. This is more complicated since we're defining a ring (in fact, an algebra!) structure. 

Note that `rel` is not an equivalence relation, so I used `quot` instead of `quotient`. I could have used `eqv_gen`, but I couldn't see any benefit in doing this -- it would have just added more typing.

adamtopaz replied:
Is there any reason why we would want to actually have a setoid here?

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,378 @@
+def lift {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) : tensor_algebra R M →ₐ[R] A :=
+{ to_fun := λ a, quot.lift_on a (lift_fun _ _ f) $ λ a b h,
+  begin
+    induction h,
```
I think this might be clearer with a `match` expression where you can name each field

adamtopaz replied:
I'm not sure what you mean. Can you `match` in a tactic block?

eric-wieser replied:
I think I was mistaken. What you can do is added `case zero_lin: ` etc before each `{}` block.

adamtopaz committed:
Simplify construction by changing defn of pre

jcommelin commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,350 @@
+
+universes u1 u2 u3
+
+variables (R : Type u1) [comm_ring R]
```
We have tensor products of semimodules... How much of this file works in that generality? I can imagine that secretly sometimes we might be interested in that case, as mathematicians. Even though we never really talk or write about it.
(Note: I didn't think about this at all. I suggest you just ask lean. If lots of stuff breaks in a nontrivial way... then keep it like it was.)

adamtopaz replied:
Aren't algebras only defined for commutative rings?

adamtopaz replied:
Oh, they're defined for comm_semirings, so all of this should just work in that level of generality.

jcommelin commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,350 @@
+instance : inhabited (pre R M) := ⟨of_scalar 0⟩
+
+-- These instances are only used to simplify the notation.
+@[nolint doc_blame]
```
Can you also add all the nolint attributes in one go, just as you did with the local instance attribute?

PatrickMassot replied:
I don't think we should add any `@[nolint doc_blame]` anywhere in mathlib. The possibility to add this annotation is meant only for transitioning from undocumented mathlib to documented mathlib.

adamtopaz replied:
No problem. I'll add the docstrings for these definition.

jcommelin commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,350 @@
+  pre.has_one pre.has_scalar pre.has_neg
+
+@[nolint doc_blame]
+def lift_fun {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) : pre R M → A :=
```
I think it would be good to write short docstrings for these definitions. If anyone needs to refactor this code 6 months (or 3 years) from now, they will be very happy.

adamtopaz replied:
Sure, I can do that. I was considering `lift_fun` as just an auxiliary thing needed to define `lift`, but it's no problem to add a docstring for this as well.

jcommelin reviewed (commented):
Nice work!

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,350 @@
+-- force of to be linear
+| add_lin {a b : M} : rel ↑(a+b) (↑a + ↑b)
+| smul_lin {r : R} {a : M} : rel ↑(r • a) (↑r * ↑a)
+-- force const to be a central ring morphism
```
```suggestion
-- force of_scalar to be a central ring morphism
```

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,350 @@
+| add_assoc {a b c : pre R M} : rel (a + b + c) (a + (b + c))
+| add_comm {a b : pre R M} : rel (a + b) (b + a)
+| neg_add {a : pre R M} : rel (-a + a) 0
+| zero_add {a : pre R M} : rel (0 + a) a
```
Matching `add_comm_group` would be good here:

```suggestion
-- like add_comm_group, but with rel instead of eq
| add_assoc {a b c : pre R M} : rel (a + b + c) (a + (b + c))
| add_comm {a b : pre R M} : rel (a + b) (b + a)
| add_left_neg {a : pre R M} : rel (-a + a) 0
| zero_add {a : pre R M} : rel (0 + a) a
```

adamtopaz replied:
This doesn't apply anymore, due to the ring -> semiring change.

adamtopaz committed:
Generalize to semirings

PatrickMassot commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,350 @@
+  pre.has_coe_module pre.has_coe_ring pre.has_mul pre.has_add pre.has_zero
+  pre.has_one pre.has_scalar pre.has_neg
+
+instance : ring (tensor_algebra R M) :=
```
This proof and the following one really seem to call for a small tactic, something like the `pi_instance` tactic.

ChrisHughes24 commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,350 @@
+  (lift R M f) ∘ (univ R M) = f := rfl
+
+theorem lift_unique {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A)
+  (g : tensor_algebra R M →ₐ[R] A) : g ∘ (univ R M) = f → g = lift R M f :=
```
It might be nice to have a extensionality theorem for maps out of the `tensor_algebra` as well as this. I would state the theorem something like
```lean
theorem hom_ext {A : Type u3} [ring A] [algebra R A] 
  {f g : tensor_algebra R M →ₐ[R] A} : 
  f.to_linear_map.comp (univ R M) = g.to_linear_map.comp (univ R M) → 
  f = g
```
This isn't standard mathlib style yet, but it's something I think is quite useful.

adamtopaz replied:
I like this. Adding it!

adamtopaz committed:
Added hom_ext

Also, changed function ccomp in theorems to a comp of lin maps.

adamtopaz committed:
Add docstrings

kim-em commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,372 @@
+/-- Coersion from `M` to `pre R M`. Note: Used for notation only. -/
+def has_coe_module : has_coe M (pre R M) := ⟨of⟩
+/-- Coersion from `R` to `pre R M`. Note: Used for notation only. -/
+def has_coe_semiring : has_coe R (pre R M) := ⟨of_scalar⟩
```
```suggestion
/-- Coercion from `M` to `pre R M`. Note: Used for notation only. -/
def has_coe_module : has_coe M (pre R M) := ⟨of⟩
/-- Coercion from `R` to `pre R M`. Note: Used for notation only. -/
def has_coe_semiring : has_coe R (pre R M) := ⟨of_scalar⟩
```

kim-em commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,372 @@
+-- force of_scalar to be a central semiring morphism
+| add_scalar {r s : R} : rel ↑(r + s) (↑r + ↑s)
+| mul_scalar {r s : R} : rel ↑(r * s) (↑r * ↑s)
+| central_scalar {r : R} {a : pre R M} : rel (r * a) (a * r)
```
```suggestion
-- force `of` to be linear
| add_lin {a b : M} : rel ↑(a+b) (↑a + ↑b)
| smul_lin {r : R} {a : M} : rel ↑(r • a) (↑r * ↑a)
-- force `of_scalar` to be a central semiring morphism
| add_scalar {r s : R} : rel ↑(r + s) (↑r + ↑s)
| mul_scalar {r s : R} : rel ↑(r * s) (↑r * ↑s)
| central_scalar {r : R} {a : pre R M} : rel (r * a) (a * r)
```

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,372 @@
+/--
+Given a linear map `f : M → A` where `A` is an `R`-algebra,
+`lift R M f` is the unique lift of `f` to a morphism of `R`-algebras
+`tensor_algebra R M → A`.
```
Am I right in that the resulting g satisfies `g(coe(m1)*coe(m2)) = f(m1)*f(m2)`?

adamtopaz replied:
Which map is g?

adamtopaz replied:
If you're referring to `lift R M f` then the answer is yes -- it's an algebra map, and its composition with `univ R M` is exactly `f` by the theorem `univ_comp_lift`.

But note that there is no coercion introduced from `M` to `tensor_algebra R M` (but it's easy to define as `univ R M`, if needed).

adamtopaz committed:
Update src/linear_algebra/tensor_algebra.lean

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

adamtopaz committed:
Update src/linear_algebra/tensor_algebra.lean

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

PatrickMassot commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,372 @@
+
+1. `tensor_algebra R M` is the tensor algebra itself.
+  It is endowed with an R-algebra structure.
+2. `univ R M` is the R-linear map `M → tensor_algebra R M`.
```
This name `univ` seems really confusing to me. I think we use this word only in `set.univ` or `finset.univ`.

eric-wieser replied:
This almost feels like `has_coe M (tensor_algebra R M)` - is that a reasonable use of `coe`?

adamtopaz replied:
It's important for this map to be linear, so a plain coercion is probably not enough.
But I agree with the comment about the name. This is meant to be `univ` as in `universal` -- in this case it's the universal map arising from some application of Yoneda. Another name is the "classifying map", but I guess `class` is even worse than `univ`. Do you have suggestions for another name?

kim-em replied:
This goes badly against my principle of hoping to have easily readable names, but often the the category theory library the universal maps are called just `ι` or `π`. (Depending roughly on whether they are into or out of the universe object.)

So this could be `tensor_algebra.ι R M`.

adamtopaz replied:
Done! Changed to \iota.

PatrickMassot commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,372 @@
+This inductive type is used to express representatives
+of the tensor algebra.
+-/
+inductive pre
```
What does `pre` stands for?

eric-wieser replied:
I think this is like `dfinsupp.pre` - a convention from who knows where for an intermediate type used only for defining a `quot` / `quotient`.

adamtopaz replied:
`pre` as in the `precursor` to the actual construction. An alternative name would be `rep` for `representative`. Do you prefer one over the other?

kim-em replied:
I think this is okay. It's a common enough prefix (ha!) in English.

PatrickMassot commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,372 @@
+  then one has `g = lift R M f`.
+-/
+
+universes u1 u2 u3
```
Why do you need those universes variables, what breaks if you put `Type*` everywhere? And why don't you use unicode indices?

adamtopaz replied:
`Type *` is probably enough in this case.

adamtopaz replied:
I tried `Type*`, and the proofs in the definition of `lift` seem to break. So I kept the universes (but with unicode subscripts this time!).

PatrickMassot committed:
feat(data/quot): missing quot.lift₂ and quot.map₂

PatrickMassot committed:
fix(linear_algebra/tensor_algebra): compress instance

PatrickMassot reviewed (commented):
I pushed two commits pointing out how to clean up this file. The [first commit](https://github.com/leanprover-community/mathlib/pull/3531/commits/1e50bc8e3e3ca1b2c21fa2f7006103337f18438e) has nothing to do with algebra, it's filling a hole in the `quot` API. You only need a dozen of lines from that commit, but not putting everything else would feel weird. Then the [second commit](https://github.com/leanprover-community/mathlib/pull/3531/commits/ac48b0aebbb0dbcc5c112cd0d2fa2c6adb71bfc8) brings the `semiring` instance to reasonable size (although I still think a tactic would be nice here). Note that the new instance is defeq to yours.

eric-wieser commented on src/data/quot.lean:
```diff
@@ -56,6 +56,66 @@ quot.lift (quot.mk s) (λ x y rxy, quot.sound (h x y rxy))
+
+/-- Descends a function `f : α → β → γ` to quotients of `α` and `β`. -/
+attribute [reducible, elab_as_eliminator]
+protected def lift₂
```
Can this be used to simplify `quotient.lift₂`?

PatrickMassot replied:
Yes, you could write:
```lean
def quotient.lift₂' {α β φ: Sort*} [s₁ : setoid α] [s₂ : setoid β] 
  (f : α → β → φ) (h : ∀ (a₁ : α) (a₂ : β) (b₁ : α) (b₂ : β),
       a₁ ≈ b₁ → a₂ ≈ b₂ → f a₁ a₂ = f b₁ b₂) :
    quotient s₁ → quotient s₂ → φ :=
quot.lift₂ f (λ a b₁ b₂ hb, h a b₁ a b₂ (setoid.refl a) hb)
             (λ a₁ a₂ b ha, h a₁ b a₂ b ha (setoid.refl b))
```
but that change would need to happen in the core library and would probably break of lot of things.

eric-wieser replied:
I suppose the follow up question is, would that ever be worth it?

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -132,115 +132,25 @@ local attribute [instance]
-    apply rel.mul_assoc,
+    rw [quot.sound rel.add_comm, quot.sound rel.zero_add],
   end,
+  add_comm := by { rintros ⟨⟩ ⟨⟩, exact quot.sound rel.add_comm },
```
I've never seen `⟨⟩` used with no contents in rintros - does that just mean "make up a local name"?

PatrickMassot replied:
Yes, it makes up a local name that we won't use, but the important part is not the fact that we don't provide a name, it's the special support for `quot` in `rintros` that replaces the `quot.exists_rep` dance.

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,282 @@
+  (lift R M f).to_linear_map.comp (univ R M) = f := by {ext, refl}
+
+theorem lift_unique {A : Type u3} [semiring A] [algebra R A] (f : M →ₗ[R] A)
+  (g : tensor_algebra R M →ₐ[R] A) : g.to_linear_map.comp (univ R M) = f → g = lift R M f :=
```
Would it make sense to combine these into a single `iff`?
```suggestion
theorem lift_unique {A : Type u3} [semiring A] [algebra R A] (f : M →ₗ[R] A)
  (g : tensor_algebra R M →ₐ[R] A) : g.to_linear_map.comp (univ R M) = f ↔ g = lift R M f :=
```
obviously with the necessary changes to the proof below

adamtopaz replied:
Mathematically, yes, these two theorems are just a single logical equivalence. But my guess is that in practice both theorems are useful as they stand.

On a related note: should `univ_comp_lift` be tagged with `simp`?

kim-em replied:
Yes, it should be a simp lemma.

kim-em replied:
One possible advantage of making `lift_unique` an `iff` is that it could then also be used in `simp` and `rw`. (And possibly even be marked with `@[simp]`.

adamtopaz replied:
Yes that's a good point. But I still think it's worthwhile to keep `univ_comp_lift` around, since I can envision a situation where I would want simp to simplify a bunch of compositions like those which appear in that theorem.

adamtopaz replied:
Done!

adamtopaz committed:
Some cleanup

1. Some proofs shortened using the rintros trick
2. Renamed `univ` to `canon`.
3. Unicode subscripts for universe variables.

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,248 @@
+  left_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.left_distrib },
+  right_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.right_distrib },
+  zero_mul := by { rintros ⟨⟩, exact quot.sound rel.zero_mul },
+  mul_zero := by { rintros ⟨⟩, exact quot.sound rel.mul_zero } }
```
Can and should this be further shortened to the following?
```suggestion
  mul_zero := λ ⟨⟩, quot.sound rel.mul_zero }
```

adamtopaz replied:
I don't think this works, but I could be wrong. Did you try it?

adamtopaz replied:
Yeah, it doesn't work. Looks like this sort of pattern matching doesn't work with `quot`.

adamtopaz committed:
Some more cleanup

1. Rename `canon` to `\iota`.
2. Add simp tags
3. Make `lift_unique` an iff.

adamtopaz committed:
Fix typo in documentation.

adamtopaz committed:
Even more cleanup.

kim-em commented on src/data/quot.lean:
```diff
@@ -56,6 +56,66 @@ quot.lift (quot.mk s) (λ x y rxy, quot.sound (h x y rxy))
+   (f : α → β → γ)
+   (hr : ∀ a b₁ b₂, s b₁ b₂ → f a b₁ = f a b₂)
+   (hs : ∀ a₁ a₂ b, r a₁ a₂ → f a₁ b = f a₂ b)
+   (q₁ : quot r) (q₂ : quot s) : γ :=
```
```suggestion
  (f : α → β → γ)
  (hr : ∀ a b₁ b₂, s b₁ b₂ → f a b₁ = f a b₂)
  (hs : ∀ a₁ a₂ b, r a₁ a₂ → f a₁ b = f a₂ b)
  (q₁ : quot r) (q₂ : quot s) : γ :=
```

kim-em commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+    induction h,
+    { change f _ = f _ + f _, simp, },
+    { change f _ = (algebra_map _ _ _) * f _, rw linear_map.map_smul,
+      exact algebra.smul_def h_r (f h_a) },
```
```suggestion
    { change f _ = (algebra_map _ _ _) * f _, 
      rw linear_map.map_smul,
      exact algebra.smul_def h_r (f h_a) },
```
etc. (Although I admit I'm sloppy about this; if it really feels like two tactics are doing one thing, I guess on a single line is fine.)

kim-em commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+    { change (algebra_map _ _ _) * _ = algebra_map _ _ _, simp },
+    { change _ * (algebra_map _ _ _) = algebra_map _ _ _, simp },
+  end,
+  map_one' := by change algebra_map _ _ _ = _; simp,
```
```suggestion
  map_one' := by { change algebra_map _ _ _ = _, simp, }
```
Only use `;` if you're actually using it, not just to save on characters.

kim-em commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+  ext,
+  rcases quot.exists_rep x with ⟨x,rfl⟩,
+  induction x,
+  { change (g.to_linear_map.comp (ι R M)) _ = _, rw hyp, refl },
```
```suggestion
  { change (g.to_linear_map.comp (ι R M)) _ = _, 
    rw hyp, 
    refl },
```

kim-em commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+  { exact alg_hom.commutes g x },
+  repeat { change g (quot.mk _ _ + quot.mk _ _) = _ <|> change g (quot.mk _ _ * quot.mk _ _) = _,
+    rw alg_hom.map_add <|> rw alg_hom.map_mul,
+    rw [x_ih_a, x_ih_a_1], refl },
```
I don't much like this `repeat`. I have to read carefully to even guess how many times it is repeating, and once I work it out I start to conclude that not many characters were saved by using the `repeat`!

kim-em commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+theorem hom_ext {A : Type u₃} [semiring A] [algebra R A] {f g : tensor_algebra R M →ₐ[R] A} :
+  f.to_linear_map.comp (ι R M) = g.to_linear_map.comp (ι R M) → f = g := λ hyp,
+  let h := g.to_linear_map.comp (ι R M) in
+  by rw [show g = lift R M h, by rw ←lift_unique, ←lift_unique, hyp]
```
New line after the `:=`, and then no indenting required in the proof.

I'm not so keen on the `show ..., by ...` inside a `rw`, but I guess people do it. I would just `have` it first then `rw` by it.

kim-em commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+    { change _ * (_ + _) = _ * _ + _ * _, rw left_distrib, },
+    { change (_ + _) * _ = _ * _ + _ * _, rw right_distrib, },
+    repeat { set G := lift_fun R M f,
+      change G _ + G _ = G _ + G _ <|> change G _ * G _ = G _ * G _, rw h_ih, },
```
Not so keen on this `repeat` here, given you have to use `<|>` inside.

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+begin
+  refine ⟨λ hyp, _, λ hyp, by simp [hyp]⟩,
+  ext,
+  rcases quot.exists_rep x with ⟨x,rfl⟩,
```
```suggestion
  rcases x,
```

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+theorem lift_unique {A : Type u₃} [semiring A] [algebra R A] (f : M →ₗ[R] A)
+  (g : tensor_algebra R M →ₐ[R] A) : g.to_linear_map.comp (ι R M) = f ↔ g = lift R M f :=
+begin
+  refine ⟨λ hyp, _, λ hyp, by simp [hyp]⟩,
```
Clearer without `simp`, maybe?
```suggestion
  refine ⟨λ hyp, _, λ hyp, by {rw hyp, exact ι_comp_lift f}⟩,
```

eric-wieser commented on src/linear_algebra/tensor_algebra.lean:
```diff
@@ -0,0 +1,239 @@
+
+@[simp]
+theorem ι_comp_lift {A : Type u₃} [semiring A] [algebra R A] (f : M →ₗ[R] A) :
+  (lift R M f).to_linear_map.comp (ι R M) = f := by {ext, refl}
```
Is this similar theorem also useful?
```lean
@[simp]
theorem lift_comp_ι {A : Type u₃} [semiring A] [algebra R A] (g : tensor_algebra R M →ₐ[R] A) :
  lift R M (g.to_linear_map.comp (ι R M)) = g :=
begin
  symmetry,
  ext,
  rcases x,
  induction x,
  case pre.of: x { refl },
  case pre.of_scalar: x { exact alg_hom.commutes g x },
  case pre.add: a b ha hb { change g (quot.mk _ _ + quot.mk _ _) = _,
    rw alg_hom.map_add,
    rw [ha, hb], refl },
  case pre.mul: a b ha hb { change g (quot.mk _ _ * quot.mk _ _) = _,
    rw alg_hom.map_mul,
    rw [ha, hb], refl },
end
```

With it, you can implement `lift_unique` as `⟨λ hyp, by simp [←hyp], λ hyp, by simp [hyp]⟩`

adamtopaz committed:
Cleanup; add lift_comp_\iota

adamtopaz committed:
Remove universe variables.

adamtopaz committed:
Add implementation details

adamtopaz committed:
Small change

adamtopaz committed:
Clean up documentation.

adamtopaz committed:
Update src/data/quot.lean

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>
